File: backend/src/main/java/com/eventflow/modules/catalog/api/dto/CatalogDtos.java
package com.eventflow.modules.catalog.api.dto;

import com.eventflow.shared.web.MoneyDto;
import com.fasterxml.jackson.annotation.JsonInclude;
import jakarta.validation.constraints.DecimalMax;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/** DTOs espejo EXACTO de components.schemas del OpenAPI congelado (tag catalog/organizer/me/admin). */
public final class CatalogDtos {

    private CatalogDtos() {
    }

    public record CategoryDto(int id, String name, String icon) {
    }

    public record SponsorDto(UUID id, String name, String logoUrl, String website) {
    }

    public record ZoneDto(UUID id, String name, int capacity) {
    }

    public record TicketTypeDto(UUID id, String name, String description, MoneyDto price,
                                String zoneName, boolean available, Instant salesEndsAt) {
    }

    public record EventPolicyPublicDto(Instant refundWindowEndsAt, boolean exchangeEnabled,
                                       int exchangeDepreciationPct, boolean waitlistEnabled,
                                       int qrVisibilityHoursBefore) {
    }

    /** isFavorite solo presente autenticado (schema: [boolean, null]). */
    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record EventSummaryDto(UUID id, String title, String venueName, Instant startsAt, Instant endsAt,
                                  String timezone, String status, String coverUrl, CategoryDto category,
                                  MoneyDto priceFrom, Boolean isFavorite) {
    }

    public record OrganizerDto(UUID id, String name) {
    }

    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record EventDetailDto(UUID id, String title, String venueName, Instant startsAt, Instant endsAt,
                                 String timezone, String status, String coverUrl, CategoryDto category,
                                 MoneyDto priceFrom, Boolean isFavorite, String description, String address,
                                 Double latitude, Double longitude, OrganizerDto organizer,
                                 List<TicketTypeDto> ticketTypes, List<ZoneDto> zones, List<Object> parkings,
                                 List<SponsorDto> sponsors, EventPolicyPublicDto policies, boolean waitlistOpen) {
    }

    public record CreateEventRequest(
            @NotBlank @Size(min = 3, max = 200) String title,
            @NotNull String description,
            @NotNull Integer categoryId,
            @NotBlank String venueName,
            String address,
            @DecimalMin(value = "-90") @DecimalMax(value = "90") Double latitude,
            @DecimalMin(value = "-180") @DecimalMax(value = "180") Double longitude,
            @NotBlank String timezone,
            @NotNull Instant startsAt,
            @NotNull Instant endsAt) {
    }

    public record UpdateEventRequest(
            @Size(min = 3, max = 200) String title,
            String description,
            Integer categoryId,
            String venueName,
            String address,
            @DecimalMin(value = "-90") @DecimalMax(value = "90") Double latitude,
            @DecimalMin(value = "-180") @DecimalMax(value = "180") Double longitude,
            String timezone,
            Instant startsAt,
            Instant endsAt,
            String coverUrl) {
    }

    public record EventPolicyRequest(
            Instant refundWindowEndsAt,
            @NotNull Boolean exchangeEnabled,
            @NotNull @Min(0) @Max(100) Integer exchangeDepreciationPct,
            Instant exchangeListingDeadline,
            @NotNull Boolean waitlistEnabled,
            @NotNull @Min(1) Integer waitlistOfferMinutes,
            @NotNull @Min(1) Integer tempReservationMinutes,
            @NotNull @Min(0) Integer qrVisibilityHoursBefore,
            @NotNull @Min(1) Integer qrExpirationMinutes,
            String cancellationPolicy,
            Map<String, Object> extraPolicies) {
    }

    public record EventPolicyResponseDto(UUID eventId, Instant refundWindowEndsAt, boolean exchangeEnabled,
                                         int exchangeDepreciationPct, Instant exchangeListingDeadline,
                                         boolean waitlistEnabled, int waitlistOfferMinutes,
                                         int tempReservationMinutes, int qrVisibilityHoursBefore,
                                         int qrExpirationMinutes, String cancellationPolicy,
                                         Map<String, Object> extraPolicies, int version) {
    }

    public record CreateZoneRequest(@NotBlank @Size(max = 80) String name, @NotNull @Min(1) Integer capacity) {
    }

    public record CategoryRequest(@NotBlank @Size(max = 80) String name, String icon, @NotNull Boolean active) {
    }

    public record SponsorRequest(@NotBlank @Size(max = 120) String name, String logoUrl, String website) {
    }
}

File: app/src/main/java/com/app/eventflow/data/remote/dto/catalog/CatalogDtos.kt
package com.app.eventflow.data.remote.dto.catalog

import kotlinx.serialization.Serializable

/** DTOs espejo EXACTO de components.schemas del OpenAPI congelado (tag catalog/me). */

@Serializable
data class MoneyDto(val amount: String, val currency: String)

@Serializable
data class CategoryDto(val id: Int, val name: String, val icon: String? = null)

@Serializable
data class SponsorDto(val id: String, val name: String, val logoUrl: String? = null, val website: String? = null)

@Serializable
data class ZoneDto(val id: String, val name: String, val capacity: Int)

@Serializable
data class TicketTypeDto(
    val id: String,
    val name: String,
    val description: String? = null,
    val price: MoneyDto,
    val zoneName: String? = null,
    val available: Boolean,
    val salesEndsAt: String? = null,
)

@Serializable
data class EventPolicyPublicDto(
    val refundWindowEndsAt: String? = null,
    val exchangeEnabled: Boolean,
    val exchangeDepreciationPct: Int,
    val waitlistEnabled: Boolean,
    val qrVisibilityHoursBefore: Int,
)

@Serializable
data class OrganizerDto(val id: String, val name: String)

@Serializable
data class EventSummaryDto(
    val id: String,
    val title: String,
    val venueName: String,
    val startsAt: String,
    val endsAt: String,
    val timezone: String,
    val status: String,
    val coverUrl: String? = null,
    val category: CategoryDto,
    val priceFrom: MoneyDto? = null,
    val isFavorite: Boolean? = null,
)

@Serializable
data class EventDetailDto(
    val id: String,
    val title: String,
    val venueName: String,
    val startsAt: String,
    val endsAt: String,
    val timezone: String,
    val status: String,
    val coverUrl: String? = null,
    val category: CategoryDto,
    val priceFrom: MoneyDto? = null,
    val isFavorite: Boolean? = null,
    val description: String,
    val address: String? = null,
    val latitude: Double? = null,
    val longitude: Double? = null,
    val organizer: OrganizerDto,
    val ticketTypes: List<TicketTypeDto> = emptyList(),
    val zones: List<ZoneDto> = emptyList(),
    val sponsors: List<SponsorDto> = emptyList(),
    val policies: EventPolicyPublicDto,
    val waitlistOpen: Boolean,
)

@Serializable
data class CursorMetaDto(val hasNext: Boolean, val nextCursor: String? = null)

@Serializable
data class EventsPageDto(val data: List<EventSummaryDto>, val meta: CursorMetaDto)
